### PR TITLE
HGCal trigger updates

### PR DESCRIPTION
**Pull request by jbsauvan**
Main changes:
* Add ntuplizer modules
* Add a TriggerCell data format
* Add new type of trigger cell selection in the front-end
* Add trigger cell energy calibration
* Produce the HGCal trigger geometry with an ESProducer
* Update the trigger module mapping for the V7 and V8 HGCal geometry. Depends on external cms-data/L1Trigger-L1THGCal#3, but tests can be run without these data (HGCal trigger is currently commented out in SimL1Emulator_cff)

### CODE REVIEW

**Comment by jbsauvan** [comment]
Please test

**Comment by rekovic** [comment]
please test

**Comment by smuzaffar** [comment]
please test with cms-sw/cmsdist#2787

File: L1Trigger/L1THGCal/plugins/be_algorithms/FullModuleSumAlgo.cc [hunk 1]
```diff
@@ -1,27 +1,56 @@
...
+
+    protected:
+        using Algorithm<FECODEC>::codec_;
+
+    public:    
         FullModuleSumAlgo(const edm::ParameterSet& conf):
-            Algorithm<HGCalBestChoiceCodec>(conf),
-            cluster_product_( new l1t::HGCalClusterBxCollection ){}
+        Algorithm<FECODEC>(conf),
+        cluster_product_( new l1t::HGCalClusterBxCollection ){}
```
Comment (rekovic) [inline L1Trigger/L1THGCal/plugins/be_algorithms/FullModuleSumAlgo.cc]: How is this memory deallocated? Can you use smart pointers?

File: L1Trigger/L1THGCal/test/HGCalTriggerBestChoiceTester.cc [hunk 2]
```diff
@@ -105,12 +105,6 @@ HGCalTriggerBestChoiceTester::HGCalTriggerBestChoiceTester(const edm::ParameterS
...
 {
-    //setup geometry 
-    const edm::ParameterSet& geometryConfig = conf.getParameterSet("TriggerGeometry");
-    const std::string& trigGeomName = geometryConfig.getParameter<std::string>("TriggerGeometryName");
-    HGCalTriggerGeometryBase* geometry = HGCalTriggerGeometryFactory::get()->create(trigGeomName,geometryConfig);
-    triggerGeometry_.reset(geometry);
-
     //setup FE codec
     const edm::ParameterSet& feCodecConfig = conf.getParameterSet("FECodec");
     codec_.reset( new HGCalBestChoiceCodecImpl(feCodecConfig) );
```
Comment (rekovic) [inline L1Trigger/L1THGCal/test/HGCalTriggerBestChoiceTester.cc]: Same....How is this memory deallocated? Use smart pointers?

File: L1Trigger/L1THGCal/plugins/be_algorithms/SingleCellClusterAlgo.cc [hunk 3]
```diff
@@ -1,76 +1,100 @@
...
+
+    protected:
+        using Algorithm<FECODEC>::codec_;
 
+    public:
         SingleCellClusterAlgo(const edm::ParameterSet& conf):
-            Algorithm<HGCalBestChoiceCodec>(conf),
-            cluster_product_( new l1t::HGCalClusterBxCollection ){}
+            Algorithm<FECODEC>(conf),
+            cluster_product_( new l1t::HGCalTriggerCellBxCollection ),
```
Comment (rekovic) [inline L1Trigger/L1THGCal/plugins/be_algorithms/SingleCellClusterAlgo.cc]: same...How is this memory deallocated? Can you use smart pointers?

File: L1Trigger/L1THGCal/plugins/be_algorithms/SingleCellClusterAlgo.cc [hunk 4]
```diff
@@ -1,76 +1,100 @@
...
+ 
         virtual void putInEvent(edm::Event& evt) override final 
         {
             evt.put(std::move(cluster_product_),name());
         }
 
         virtual void reset() override final 
         {
-            cluster_product_.reset( new l1t::HGCalClusterBxCollection );
+            cluster_product_.reset( new l1t::HGCalTriggerCellBxCollection );
```
Comment (rekovic) [inline L1Trigger/L1THGCal/plugins/be_algorithms/SingleCellClusterAlgo.cc]: same..How is this memory deallocated? Can you use smart pointers?

**Review by rekovic — COMMENTED** [review]
Issues with "new" statements need to be addressed.

File: L1Trigger/L1THGCal/plugins/be_algorithms/FullModuleSumAlgo.cc (hunk 1 above)
Comment (jbsauvan) [inline L1Trigger/L1THGCal/plugins/be_algorithms/FullModuleSumAlgo.cc]: `cluster_product_` is already a `unique_ptr`

File: L1Trigger/L1THGCal/plugins/be_algorithms/SingleCellClusterAlgo.cc (hunk 3 above)
Comment (jbsauvan) [inline L1Trigger/L1THGCal/plugins/be_algorithms/SingleCellClusterAlgo.cc]: `cluster_product_` is already a `unique_ptr`

File: L1Trigger/L1THGCal/plugins/be_algorithms/SingleCellClusterAlgo.cc (hunk 4 above)
Comment (jbsauvan) [inline L1Trigger/L1THGCal/plugins/be_algorithms/SingleCellClusterAlgo.cc]: `cluster_product_` is already a `unique_ptr`

File: L1Trigger/L1THGCal/test/HGCalTriggerBestChoiceTester.cc (hunk 2 above)
Comment (jbsauvan) [inline L1Trigger/L1THGCal/test/HGCalTriggerBestChoiceTester.cc]: `codec_` is a `unique_ptr`

**Comment by jbsauvan** [comment]
Hi @rekovic, is the memory management fine? Is there something else that needs to be changed / improved?
Thanks!

**Comment by jbsauvan** [comment]
Hi @rekovic 
Could you consider signing in case there is no more issue? Or please tell me what should be improved.
Thanks

**Comment by kpedro88** [comment]
+1

**Comment by jbsauvan** [comment]
Hi @rekovic, any chance of having further comments or L1T sign-off for this PR soon?

**Comment by kpedro88** [comment]
@rekovic, @mulhearn ping

**Comment by rekovic** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**